### ISSUE TITLE
ValueError: Related model 'wagtailcore.pagerevision' cannot be resolved

### ISSUE THREAD

**Issue by kmtracey** [wagtail/wagtail-localize#642]
Adding wagtail 4.0.2 and wagtail-localize 1.3.1 to an existing project the first attempt to migrate fails 

```
... lots of wagtailcore migrations...
Applying wagtail_localize.0001_initial...Traceback (most recent call last):
... lots of traceback...
... django/db/models/fields/related.py", line 712, in resolve_related_fields
    raise ValueError(
ValueError: Related model 'wagtailcore.pagerevision' cannot be resolved
```

Before the failure  wagtailcore through `wagtailcore.0077_alter_revision_user` had been run. It seems, however, that the `get_revision_model()` function in wagtail_localize 0001 is called before any of the migrations that need to be run are run...that is it is called very early during migration planning (?), and at that time the `wagtailcore.0070_rename_pagerevision_revision` looked for by the initial wagail_localize migration had not run...but by the time initial localize migration is actually run the core migrations have moved on past 0070.

Seems `get_revision_model` is making a decision based on what has already run but to be accurate needs also to know what will run in this run of "migrate" before the 0001 migration for localize?

**Comment by zerolab** [comment]
The honest answer is 🤷‍♂️
I went trough many iterations, #613, #616, #630. As well asseveral alphas releases to get as much testing as possible. Tested manually with a bunch of scenarios.

> Adding wagtail 4.0.2 and wagtail-localize 1.3.1 to an existing project the first attempt to migrate fails

Doesn't say a lot, I'm afraid. What are the steps to reproduce? Were you able to reproduce this on other projects?

**Comment by laymonage** [comment]
Hit this while working on https://github.com/wagtail/guide as well. I installed wagtail-localize on the project a few days ago (after upgrading to 4.0.4), and the weird thing is I didn't encounter this after the installation, but now I'm getting this today and I don't think I've changed anything at all that would cause this...

I'll do some more digging and report back if I find anything.

**Comment by laymonage** [comment]
Okay, I think the issue happens when you try to run Wagtail's 4.0 migrations **and** wagtail-localize's migrations in the same `migrate` run.

The `get_revision_model()` function is run at module initialisation time, because it's within the `Migration` class definition:

https://github.com/wagtail/wagtail-localize/blob/fb4db60814a1362e93787d84a5a9dbaf4bb37f71/wagtail_localize/migrations/0001_initial.py#L167

If Wagtail's `0070_rename_pagerevision_revision` migration is run at the same `migrate` run as wagtail-localize's `0001_initial` revision, then the `0070_rename_pagerevision_revision` will not exist in the `MigrationRecorder.Migration` model at the time the `0001_initial` revision is loaded into memory, so `get_revision_model()` returns `wagtailcore.PageRevision`. However, the `0070_rename_pagerevision_revision` is run *before* wagtail-localize's `0001_initial`, so by the time `0001_initial` is run, the model state has been renamed to `wagtailcore.Revision`, hence `ValueError: Related model 'wagtailcore.pagerevision' cannot be resolved`.

Not sure what the solution is, I'm assuming the `MigrationRecorder.Migration` check exists for a reason (if not we could probably remove it). We can probably move the `operations` definition inside the `__init__()` so that `get_revision_model()` is executed later?

**Comment by zerolab** [comment]
> Okay, I think the issue happens when you try to run Wagtail's 4.0 migrations and wagtail-localize's migrations in the same migrate run.

This was one of the scenarios I tried locally for #630 🤔

**Comment by laymonage** [comment]
I'm thinking that we could probably combine the solutions in #613 and #630. However, we probably need to add an additional check before updating the `core_migration` to `0076_modellogentry_revision` to check whether the `0001_initial` migration itself has been run or not.

If I understand correctly, for new installations with Wagtail>=4.0, we'd like to make the migrations point to `wagtailcore.Revision` and make it depend on Wagtail 4.0's last migration.

However, #615 happens because `0001_initial` has already been run with a previous version of wagtail-localize and Wagtail (pre 3.0). I'm assuming for such installations, the `0001_initial` migration should depend on pre-Wagtail 4.0 migration, so it should still point to `0057_page_locale_fields_notnull`.

This might not completely solve the problem, though, I still need to dig deeper. I wonder if there's a way to find out the order of migrations that have been applied to the database and figure out the upgrade pattern and migration strategy from there...

**Comment by laymonage** [comment]
> Okay, I think the issue happens when you try to run Wagtail's 4.0 migrations and wagtail-localize's migrations in the same migrate run.

In addition to the above, it seems that this happens when for some reason the resolved migration order decided by Django is to run `0001_initial` before `0070_rename_pagerevision_revision`. This can happen because the constraints of `0001_initial` allow for it to be run both before and after the `0070_rename_pagerevision_revision`.

I don't know what I did with https://github.com/wagtail/guide/pull/175 that triggers this, but you can check out that PR branch for repro.

So I think the solution is to ensure that:
- if `0070_rename_pagerevision_revision` hasn't been run, make sure `0001_initial` is run before that.
- otherwise, we don't really care whether `0001_initial` is run before or after `0070_rename_pagerevision_revision`, but make sure that the `revision_model` points to the renamed model.

Not sure if this approach is sound, but this seems to work for me:

```
def get_revision_model_and_run_before():
    revision_model = "wagtailcore.PageRevision"
    run_before = [("wagtailcore", "0070_rename_pagerevision_revision")]
    if WAGTAIL_VERSION >= (4, 0, 0):
        try:
            if MigrationRecorder.Migration.objects.filter(
                app="wagtailcore", name="0070_rename_pagerevision_revision"
            ).exists():
                revision_model = "wagtailcore.Revision"
                run_before = []
        except (django.db.utils.OperationalError, django.db.utils.ProgrammingError):
            # normally happens when running tests
            pass

    return revision_model, run_before


revision_model, run_before = get_revision_model_and_run_before()


class Migration(migrations.Migration):
    initial = True

    dependencies = [
        ("wagtailcore", "0057_page_locale_fields_notnull"),
        ("contenttypes", "0002_remove_content_type_name"),
    ]

    run_before = run_before

    operations = [
        ...,
        migrations.CreateModel(
            name="TranslationLog",
            fields=[
                (
                    "id",
                    models.AutoField(
                        auto_created=True,
                        primary_key=True,
                        serialize=False,
                        verbose_name="ID",
                    ),
                ),
                ("created_at", models.DateTimeField(auto_now_add=True)),
                (
                    "locale",
                    models.ForeignKey(
                        on_delete=django.db.models.deletion.CASCADE,
                        related_name="translation_logs",
                        to="wagtailcore.Locale",
                    ),
                ),
                (
                    "page_revision",
                    models.ForeignKey(
                        blank=True,
                        null=True,
                        on_delete=django.db.models.deletion.SET_NULL,
                        related_name="+",
                        to=revision_model,
                    ),
                ),
                (
                    "source",
                    models.ForeignKey(
                        on_delete=django.db.models.deletion.CASCADE,
                        related_name="translation_logs",
                        to="wagtail_localize.TranslationSource",
                    ),
                ),
            ],
        ),
        ...,
    ]
```

**Comment by laymonage** [comment]
Oh, let me make a minor fix:

```py
def get_revision_model_and_run_before():
    revision_model = "wagtailcore.PageRevision"
    run_before = []
    if WAGTAIL_VERSION >= (4, 0, 0):
        run_before = [("wagtailcore", "0070_rename_pagerevision_revision")]
        try:
            if MigrationRecorder.Migration.objects.filter(
                app="wagtailcore", name="0070_rename_pagerevision_revision"
            ).exists():
                revision_model = "wagtailcore.Revision"
                run_before = []
        except (django.db.utils.OperationalError, django.db.utils.ProgrammingError):
            # normally happens when running tests
            pass

    return revision_model, run_before


revision_model, run_before = get_revision_model_and_run_before()
```

This ensures that:
  - for Wagtail<4.0, `0001_initial` works the way it has always been.
  - for Wagtail>=4.0:
    - if `0070_rename_pagerevision_revision` **hasn't** been run at the start of the `migrate`:
      - run `0001_initial` before `0070_rename_pagerevision_revision`. At this point, `0001_initial` should still point to `PageRevision`.
    - if `0070_rename_pagerevision_revision` **has** been run at the start of the `migrate`:
      - we don't care if `0001_initial` was run before or after, but make sure to use the new `Revision` name.

I *think* it should work...

**Comment by kmtracey** [comment]
FWIW this last one works for my case of simultaneously adding wagtail 4.0.2 + wagtail_localize 1.3.1(+) to a project (where I hit the `cannot be resolved` issue). I also have another project with wagtail 2.15.1 and on that one I was able to update wagtail_localize to 1.3.1(+) (from 1.0) and `migrate` ran without error (finding nothing to do).

**Comment by zerolab** [comment]
@sage care to submit a PR?
@kmtracey thank you for testing this

**Comment by zerolab** [comment]
Fixed in v1.3.2 
Thank you @kmtracey for the report and testing, @laymonage for the fix
